### PR TITLE
Feature/complex polygon

### PR DESCRIPTION
**Pull request by TrevorBurgoyne**
# Complex Polygons

## Description

- Added click-and-drag functionality when drawing polygons or polylines.
- Changed polygon representation from that of a simple polygon to a complex polygon. This allows for polygons to contain holes and/or multiple disjoint regions.
    - NOTE: Self-intersecting polygons are not supported
- Hold `shift` when closing a polygon to continue annotating a new region or hole.
- Hold `shift` when moving the cursor inside a polygon to begin annotating a new region or hole.
- Press `Escape` or `crtl+z` to cancel the start of a new region or hole.
- Added fill to polygons by default, excluding any holes in the polygon.
- Added check to ensure that deprecated simple polygons are auto converted to complex polygons when loaded by ULabel.

## PR Checklist

- [x] Merged latest main
- [x] Version number in `package.json` has been bumped since last release
- [x] Version numbers match between package `package.json` and `src/version.js`
- [x] Ran `npm install` and `npm run build` AFTER bumping the version number
- [x] Updated documentation if necessary (currently just in `api_spec.md`)
- [x] Added changes to `changelog.md`

## Breaking API Changes
Polygon `spatial_payload` arrays are always Complex Polygon arrays, which is to say that instead of being an Array of points (aka a simple polygon), the `spatial_payload` is an Array of arrays of points (an array of simple polygons, which together make a complex polygon). 

This means that the previous simple polygon format is now deprecated. However, old annotations will be auto converted upon being loaded into an instance of ULabel running on the latest version. Just note that polygon annotations will be saved in this format, so any use of the labels external to ULabel will have to accommodate this change.

Added `@turf/turf`, `polygon-clipping` as npm dependencies for some polygon operations.

### CODE REVIEW

**Review by csolbs24 — COMMENTED** [review]
**Bugs:**
The edit_suggestions dialog is currently broken when using the new complex polygons. It currently works as expected while mounted to an annotation's corners, and does not behave while editing a single segment of a polygon. If you try to edit a segment in such a way, visually nothing will happen. Then if you press ctrl + z to undo the un-rendered edit, it will delete one side of the polygon. From there using the edit_suggestion dialog, on one of the corners which used to belong to the deleted polygon segment, will cause the two segments which should not be connected to connect to each other. Pressing ctrl + z will not undo this action.

**Code Suggestions:**
Throughout most of the modified code in this pr, whenever you declare a new variable in a Typescript file, it has not been typed. It is recommended to do so as that's the main benefit of using Typescript in the first place. I do realize that throughout the code base we have been quite sloppy in typing every variable appropriately, however we should try to uphold the standard whenever we can.
Additionally I've noticed that you always use the `==` equality operator instead of the `===` equality operator. `==` will try to convert values before comparing them which can often result in weird bugs. For example: 
`null == undefined` -> true
`null === undefined` -> false
`"5" == 5` -> true
`"5" === 5` -> false
In short `===` will only ever return true if the two values it's comparing are of the same type so its recommended to use it instead.

**Comment by TrevorBurgoyne** [comment]
> **Bugs:** The edit_suggestions dialog is currently broken when using the new complex polygons. It currently works as expected while mounted to an annotation's corners, and does not behave while editing a single segment of a polygon. If you try to edit a segment in such a way, visually nothing will happen. Then if you press ctrl + z to undo the un-rendered edit, it will delete one side of the polygon. From there using the edit_suggestion dialog, on one of the corners which used to belong to the deleted polygon segment, will cause the two segments which should not be connected to connect to each other. Pressing ctrl + z will not undo this action.

Good catch, I had totally forgot that was a feature. Fix was simple: [58c4d71](https://github.com/SenteraLLC/ulabel/pull/129/commits/58c4d715fae341b06247673d94aec3da73f5b578)
 
> **Code Suggestions:** Throughout most of the modified code in this pr, whenever you declare a new variable in a Typescript file, it has not been typed. It is recommended to do so as that's the main benefit of using Typescript in the first place. I do realize that throughout the code base we have been quite sloppy in typing every variable appropriately, however we should try to uphold the standard whenever we can. Additionally I've noticed that you always use the `==` equality operator instead of the `===` equality operator. `==` will try to convert values before comparing them which can often result in weird bugs. For example: `null == undefined` -> true `null === undefined` -> false `"5" == 5` -> true `"5" === 5` -> false In short `===` will only ever return true if the two values it's comparing are of the same type so its recommended to use it instead.

Equality changes: [2153662](https://github.com/SenteraLLC/ulabel/pull/129/commits/215366276216ff9c7e8a6ce079090a59c5187359)

Better typing (i think?): [ce48afd](https://github.com/SenteraLLC/ulabel/pull/129/commits/ce48afd4b2dcd0e89ab59aa0b916570943638d33)

File: src/index.js [hunk 1]
```diff
@@ -1750,14 +1783,57 @@ export class ULabel {
...
+            // If polygon is closed, fill it
+            if (spatial_type === "polygon" && GeometricUtils.is_polygon_closed(active_spatial_payload)) {
+                ctx.closePath();
+                ctx.globalAlpha = 0.2;
+                ctx.fill();
+                ctx.globalAlpha = 1.0;
+            }           
+        }
+
+        // For polygons, go back through and unfill the intersectoions of all the polygons
```
Comment (csolbs24) [inline src/index.js:1819]:
```suggestion
        // For polygons, go back through and unfill the intersections of all the polygons
```
Is "intersectoions" correct? It was changed to "intersections" in [7b2e4f1](https://github.com/SenteraLLC/ulabel/pull/129/commits/7b2e4f14cb45bbd66e4e16ce77408328c32731ef) and then reverted back to "intersectoions" in [2153662](https://github.com/SenteraLLC/ulabel/pull/129/commits/215366276216ff9c7e8a6ce079090a59c5187359)

File: src/geometric_utils.ts [hunk 2]
```diff
@@ -86,23 +87,23 @@ export class GeometricUtils {
 
     // Return the point on a polygon that's closest to a reference along with its distance
     public static get_nearest_point_on_polygon(ref_x: number, ref_y: number, spatial_payload: ULabelSpatialPayload2D, dstmax: number = Infinity, include_segments: boolean = false): object {
-        const poly_pts = spatial_payload;
+        const poly_pts: ULabelSpatialPayload2D = spatial_payload;
 
         // Initialize return value to null object
-        var ret = {
+        let ret: Object = {
```
Comment (csolbs24) [inline src/geometric_utils.ts:93]:
```suggestion
        let ret: {
            "access": number,
            "distance": number,
            "point": [number, number]
        } = {
```

**Review by csolbs24 — COMMENTED** [review]
**BUGS:**
1. Sometime between my previous review and this review `demo/resume-from.html` no longer loads any annotation when opening. I checked and `demo/row-filtering-example.html` does load annotations when opening.
2. It seems like its impossible to create a complex polygon with a "ring" of un-shaded area. By that I mean that its impossible to create a complex polygon three layers deep. As a user I would expect that if I made another polygon region inside of an un-shaded polygon region it would then become shaded again, but that does not appear to be the current behavior.

**Code:**
Thank you for typing all of you Typescript variables. In Javascript everything that is not a primitive is considered an Object. It's much more useful to type each object by what its properties are and what type each of those properties is. I provided a single example in a single line comment but it would be helpful to type each Object as best as you can. This way the Typescript compiler will tell you when you try to access a property that doesn't exist on an object, or when you try to set a property on an object to the wrong type.

**Comment by TrevorBurgoyne** [comment]
> **BUGS:**
> 
> 1. Sometime between my previous review and this review `demo/resume-from.html` no longer loads any annotation when opening. I checked and `demo/row-filtering-example.html` does load annotations when opening.

Nice catch, twas an issue caused when I switched "==" to "===" in some key places. Resolved in [3f7dce0](https://github.com/SenteraLLC/ulabel/pull/129/commits/3f7dce07e5566e396740dd9456e0a1fec1177bb5)

> 2. It seems like its impossible to create a complex polygon with a "ring" of un-shaded area. By that I mean that its impossible to create a complex polygon three layers deep. As a user I would expect that if I made another polygon region inside of an un-shaded polygon region it would then become shaded again, but that does not appear to be the current behavior.

Indeed, let me tell you this is not a trivial feature. I tried several different approaches to handle arbitrarily deep holes, but nothing adequately handled every edge case, especially when you try and draw very weird concave shapes. For now I'm going with [this](https://github.com/SenteraLLC/ulabel/pull/129/commits/44f4a7ea4d6a433cbfbfd34be942fd7ae1057798), which handles the general cases I expect to be used in the immediate term. At the very least, the polygon outlines will draw correctly even in the remaining edge cases where the fills don't work. Even with this approach, after nesting many layers on polygons with lots of points, it was getting pretty computationally expensive to recompute on every redraw, so I've added a [cache](https://github.com/SenteraLLC/ulabel/pull/129/commits/a41f0e841c20d410e9e3037d5730abe446a94526) to speed things up when not editing a given polygon.

> **Code:** Thank you for typing all of you Typescript variables. In Javascript everything that is not a primitive is considered an Object. It's much more useful to type each object by what its properties are and what type each of those properties is. I provided a single example in a single line comment but it would be helpful to type each Object as best as you can. This way the Typescript compiler will tell you when you try to access a property that doesn't exist on an object, or when you try to set a property on an object to the wrong type.

Done in [1dbaa10](https://github.com/SenteraLLC/ulabel/pull/129/commits/1dbaa106fd1d56a551252de01fb966f9a3703143)

**Comment by TrevorBurgoyne** [comment]
> > 2. It seems like its impossible to create a complex polygon with a "ring" of un-shaded area. By that I mean that its impossible to create a complex polygon three layers deep. As a user I would expect that if I made another polygon region inside of an un-shaded polygon region it would then become shaded again, but that does not appear to be the current behavior.

> Indeed, let me tell you this is not a trivial feature. I tried several different approaches to handle arbitrarily deep holes, but nothing adequately handled every edge case, especially when you try and draw very weird concave shapes. For now I'm going with [this](https://github.com/SenteraLLC/ulabel/pull/129/commits/44f4a7ea4d6a433cbfbfd34be942fd7ae1057798), which handles the general cases I expect to be used in the immediate term. At the very least, the polygon outlines will draw correctly even in the remaining edge cases where the fills don't work. Even with this approach, after nesting many layers on polygons with lots of points, it was getting pretty computationally expensive to recompute on every redraw, so I've added a [cache](https://github.com/SenteraLLC/ulabel/pull/129/commits/a41f0e841c20d410e9e3037d5730abe446a94526) to speed things up when not editing a given polygon.

UPDATE: I've reworked the way that fills are computed to be much more robust, and take less computation as well. The idea is to determine whether a complex layer is a hole or not right after it is draw or edited, and then track that in an `spatial_payload_holes` array. To do this, we still check for intersections, but once we find a valid intersection between layers, we *merge* the non-intersecting portions into one of the layers, while the intersection remains as a new layer. The result is a much more consistent visual representation of holes/fills, and also prevents really strange shapes by just merging them into the main body when possible.

**Review by csolbs24 — COMMENTED** [review]
**Bugs:**
I found a visual bug involving undo \ redo. To replicate do the following. Create a simple complex polygon. Press ctrl + z to undo the action of closing the polygon. Press ctrl + shift + z to redo the action of closing the polygon. The final segment of the polygon will end wherever the cursor was when the redo action was performed instead of its original location. This bug is only visual, however, as moving the polygon will cause it to be rendered properly again. 

**General Comments**
We have a lot of overlap of types throughout this project and I think that it's getting pretty messy and confusing. I also recently learned that the way we handle custom types in this project is not considered very good practice. In general types should be declared in 3 different places. First is inside of index.d.ts. This file should be used for types that are useful for interfacing with the project when ulabel is being imported. And index.d.ts can also be imported from internally for each file that needs a given type. The second place is inside of a dedicated types.ts (or some other generic\preferred name) for types that are useful across multiple files in the project. They can then be imported as needed from that file. Finally types can should be declared at the top of a typescript file if they are useful inside of that file, but not needed in any other file in the project. This system can keep our type system more organized and reduce duplicated types. 
For example we currently have two types that behave exactly the same. 
`type ULabelSpatialPayload = [number, number][]` and 
`type Point2D = [number, number]`
`type ULabelSpatialPayload2D = Point2D[]`

**Conclusion**
Very cool pr, I think that the code was well done and the feature is really cool. I'd like to see the visual bug I outlined above fixed before I approve, but everything I said in the general comments is just something that I think we should keep in mind going forward but I don't think is necessary for this pr.

**Comment by TrevorBurgoyne** [comment]
> **Conclusion** Very cool pr, I think that the code was well done and the feature is really cool. I'd like to see the visual bug I outlined above fixed before I approve, but everything I said in the general comments is just something that I think we should keep in mind going forward but I don't think is necessary for this pr.

This visual bug seems to be something I resolved in #130. I also opened #131 to track the general type refactor that you've outlined here.

**Review by csolbs24 — APPROVED** [review]
LGTM 👍

File: src/geometric_utils.ts [hunk 3]
```diff
@@ -71,38 +98,99 @@ export class GeometricUtils {
...
-        // return object with info about the point
+        // Return object with info about the point
         return {
             "dst": dst,
             "prop": kprop
         };
     }
 
+    // Check if two line segments intersect
+    public static line_segments_intersect(line1: LineSegment2D, line2: LineSegment2D): boolean {
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:115]: You ever consider using a library for this type of thing? I think Turf `lineIntersect` would fit the bill nicely:
https://turfjs.org/docs/#lineIntersect

Comment (DanSchoppe) [reply src/geometric_utils.ts:115]: I guess Turf is probably explicitly geospatial, which is not desirable here, so perhaps it's not a great fit after all. I guess the question still stands though — did you investigate whether a library fits the bill for this type of geometry utility function?

File: src/geometric_utils.ts [hunk 4]
```diff
@@ -71,38 +98,99 @@ export class GeometricUtils {
...
+    }
+
+    // Check if two line segments are equal
+    public static line_segments_are_equal(line1: LineSegment2D, line2: LineSegment2D): boolean {
+        return (
+            (line1[0][0] === line2[0][0]) &&
+            (line1[0][1] === line2[0][1]) &&
+            (line1[1][0] === line2[1][0]) &&
+            (line1[1][1] === line2[1][1])
+        );
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:140]: Similar to the geometry library question above, I would have probably reached for [Lodash](https://lodash.com/docs/4.17.15#isEqual) here:
```suggestion
        return _.isEqual(line1, line2);
```

File: src/geometric_utils.ts (hunk 3 above)
Comment (DanSchoppe) [reply src/geometric_utils.ts:115]: A couple candidates, for example:
https://github.com/HarryStevens/geometric
https://github.com/mathigon/euclid.js

File: src/geometric_utils.ts [hunk 5]
```diff
@@ -126,17 +214,178 @@ export class GeometricUtils {
...
+
+    // Return the intersection of two polygons
+    public static get_polygon_intersection_single(poly1: ULabelSpatialPayload2D, poly2: ULabelSpatialPayload2D): ULabelSpatialPayload2D {
+        // Convert to turf polygons
+        try {
+            let poly1_turf = turf.polygon([poly1]);
+            let poly2_turf = turf.polygon([poly2]);
+            
+            // Find intersection
+            let intersection = turf.intersect(poly1_turf, poly2_turf);
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:254]: Ah, cool, so you are using Turf! Is it problematic that Turf is tailored for GeoJSON? I'm imagining values outside of [-360, 360] or even cases where Turf is attempting to use a WGS84 ellipsoid.

File: src/geometric_utils.ts [hunk 6]
```diff
@@ -126,17 +214,178 @@ export class GeometricUtils {
...
+                // Ensure both polygons are closed
+                if (GeometricUtils.is_polygon_closed(poly1) && GeometricUtils.is_polygon_closed(poly2)) {
+                    let intersection: ULabelSpatialPayload2D = GeometricUtils.get_polygon_intersection_single(poly1, poly2);
+                    if (intersection != null) {       
+                        // Don't add duplicate intersections
+                        let is_duplicate: boolean = false;
+                        for (let k: number = 0; k < ret.length; k++) {
+                           if (GeometricUtils.polygons_are_equal(ret[k], intersection)) {
+                               is_duplicate = true;
+                               break;
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:233]: Something like the below snippet would be a way to reduce [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) and improve readability. (Disclaimer: untested code)

```js
// getCombinations(['apple', 'banana', 'cucumber']) => [ [ 'apple', 'banana' ], [ 'apple', 'cucumber' ], [ 'banana', 'cucumber' ] ]
function getCombinations (array) {
  return _.flatMap(array, (value, index, collection) =>
    _.map(_.slice(collection, index + 1), (secondValue) => [value, secondValue])
  )
}

public static get_polygon_intersections(polygons) {
  const closedPolygons = _.filter(polygons, GeometricUtils.is_polygon_closed)

  const combinations = getCombinations(closedPolygons)

  const intersections = _.map(combinations, GeometricUtils.get_polygon_intersection_single)

  const nonNullIntersections = _.compact(intersections)

  const deduplicatedIntersections = _.uniqBy(nonNullIntersections, _.isEqual)
}
```

File: src/geometric_utils.ts [hunk 7]
```diff
@@ -126,17 +214,178 @@ export class GeometricUtils {
...
+                if (poly1[i][0] === poly2[j][0] && poly1[i][1] === poly2[j][1]) {
+                    found = true;
+                    break;
+                }
+            }
+            if (!found) {
+                return false;
+            }
+        }
+        return true;
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:296]: Similar to the above `line_segments_are_equal` suggestion:
```suggestion
    return _.isEqual(poly1, poly2)
```

File: src/geometric_utils.ts [hunk 8]
```diff
@@ -126,17 +214,178 @@ export class GeometricUtils {
...
+    }
+
+    // Check if two polygons share an edge, ie if they contain an identical line segment
+    public static polygons_share_edge(poly1: ULabelSpatialPayload2D, poly2: ULabelSpatialPayload2D): boolean {
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:300]: All of these simple `input -> output` utility functions would be great to have under unit test. `polygons_share_edge` tests could include test cases for:
- polygons that share an edge
- polygons that do not share an edge
- polygons with duplicated points

File: src/geometric_utils.ts [hunk 9]
```diff
@@ -126,17 +214,178 @@ export class GeometricUtils {
...
+        let y: number = 0;
+        for (let i: number = 0; i < poly.length; i++) {
+            x += poly[i][0];
+            y += poly[i][1];
+        }
+        return [x/poly.length, y/poly.length];
+    }
+
+    // Check if poly1 is completely within poly2 by checking if any line segment of poly1 intersects with poly2
+    public static polygon_is_within_polygon(poly1: ULabelSpatialPayload2D, poly2: ULabelSpatialPayload2D): boolean {
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:351]: Can you instead just use [turf.booleanWithin](https://turfjs.org/docs/#booleanWithin)?

File: src/geometric_utils.ts [hunk 10]
```diff
@@ -126,17 +214,178 @@ export class GeometricUtils {
...
+            let pt: Point2D = poly[i];
+            let new_x: number = center[0] + (pt[0] - center[0])*scale;
+            let new_y: number = center[1] + (pt[1] - center[1])*scale;
+            ret.push([new_x, new_y]);
+        }
+        return ret;
+    }
+
+    // Get the centroid of a polygon
+    public static get_centroid_of_polygon(poly: ULabelSpatialPayload2D): Point2D {
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:340]: Can you use https://turfjs.org/docs/#center?

File: src/geometric_utils.ts [hunk 11]
```diff
@@ -126,17 +214,178 @@ export class GeometricUtils {
...
+    public static polygon_is_within_polygon(poly1: ULabelSpatialPayload2D, poly2: ULabelSpatialPayload2D): boolean {
+        if (GeometricUtils.is_polygon_closed(poly1) && GeometricUtils.is_polygon_closed(poly2)) {
+            return turf.booleanWithin(turf.polygon([poly1]), turf.polygon([poly2]));
+        } else {
+            return false;
+        }
+    }
+
+    // Check if a point is within a polygon
+    public static point_is_within_polygon(point: Point2D, poly: ULabelSpatialPayload2D): boolean {
```
Comment (DanSchoppe) [inline src/geometric_utils.ts:360]: Another candidate for https://turfjs.org/docs/#booleanWithin?

File: src/geometric_utils.ts (hunk 9 above)
Comment (TrevorBurgoyne) [reply src/geometric_utils.ts:351]: Indeed, that's actually a change that is made in #130, see [311ca59](https://github.com/SenteraLLC/ulabel/pull/130/commits/311ca59aa1f2f2d8d4e5d8556d65d4f6b3a939f7)

File: src/geometric_utils.ts (hunk 11 above)
Comment (TrevorBurgoyne) [reply src/geometric_utils.ts:261]: yup, this one I changed in #132, see [d886be0](https://github.com/SenteraLLC/ulabel/pull/132/commits/d886be019fe15946981d59b381e3dfe3d5416563)

File: src/geometric_utils.ts (hunk 5 above)
Comment (TrevorBurgoyne) [reply src/geometric_utils.ts:206]: I haven't run into any issues thus far, and it provides a lot of nice functionality so I don't have an issue using it

File: src/geometric_utils.ts (hunk 3 above)
Comment (TrevorBurgoyne) [reply src/geometric_utils.ts:115]: A lot of these functions were first drafts and are not actually used anywhere. Removed in [ef7fda8](https://github.com/SenteraLLC/ulabel/commit/ef7fda8847ae8f262b2915dd0127f3066b0111dd)

File: src/geometric_utils.ts (hunk 7 above)
Comment (TrevorBurgoyne) [reply src/geometric_utils.ts:296]: Removed in [ef7fda8](https://github.com/SenteraLLC/ulabel/commit/ef7fda8847ae8f262b2915dd0127f3066b0111dd)

File: src/geometric_utils.ts (hunk 10 above)
Comment (TrevorBurgoyne) [reply src/geometric_utils.ts:340]: Removed in [ef7fda8](https://github.com/SenteraLLC/ulabel/commit/ef7fda8847ae8f262b2915dd0127f3066b0111dd)

File: src/geometric_utils.ts (hunk 8 above)
Comment (TrevorBurgoyne) [reply src/geometric_utils.ts:300]: Removed in [ef7fda8](https://github.com/SenteraLLC/ulabel/commit/ef7fda8847ae8f262b2915dd0127f3066b0111dd)

Comment (TrevorBurgoyne) [reply src/geometric_utils.ts:300]: the point about unit tests stands, although outside the scope of these PRs
